File: 2024/days/day09/solve.py
from inputreader.reader import InputReader
from solverbase import SolverBase


class Solver(SolverBase):
    class State:
        disk: list[int]
        blocks: list[tuple[int, int, int]]

    state: State

    def __init__(self, input_reader: InputReader) -> None:
        self.state = self.State()
        self.state.blocks = []
        free_block = False
        disk: list[int] = []
        block_id = 0
        for x in input_reader.content:
            if free_block:
                disk += [-1] * int(x)
            else:
                block_len = int(x)
                self.state.blocks.append((len(disk), block_id, block_len))
                disk += [block_id] * block_len
                block_id += 1
            free_block = not free_block
        self.state.disk = disk

    def part1(self) -> int:
        disk = self.state.disk[::]
        disklen = len(disk)
        first_free = 0
        for j, block in enumerate(disk[::-1]):
            i = disklen - j - 1

            if block < 0:
                continue
            while disk[first_free] > -1:
                first_free += 1
                if first_free >= disklen:
                    break
            else:
                if i <= first_free:
                    break
                # print(f"moving {block} from {i} to {first_free}")
                disk[first_free] = block
                disk[i] = -1
                continue
            break
        res = 0
        for i, b in enumerate(disk):
            if b < 0:
                break
            res += i * b
        return res

    def part2(self) -> int:
        disk = self.state.disk[::]
        free_list: list[list[int]] = []
        free_start = 0
        free_len = 0
        for i, b in enumerate(disk):
            if b < 0:
                if free_len == 0:
                    free_start = i
                free_len += 1
            else:
                if free_len > 0:
                    free_list.append([free_start, free_len])
                    free_len = 0

        for block_indx, block_id, block_len in self.state.blocks[::-1]:
            for freelistidx, (free_idx, free_len) in enumerate(free_list):
                if block_indx <= free_idx:
                    break
                if free_len >= block_len:
                    for i in range(block_len):
                        disk[block_indx + i] = -1
                        disk[free_idx + i] = block_id
                    if free_len > block_len:
                        free_list[freelistidx] = [
                            free_idx + block_len,
                            free_len - block_len,
                        ]
                    else:
                        free_list.pop(freelistidx)
                    break
            # for x in disk:
            #     if x < 0:
            #         print(".", end="")
            #     else:
            #         print(x, end="")
            # print()

        res = 0
        for i, b in enumerate(disk):
            if b < 0:
                continue
            res += i * b
        return res


if __name__ == "__main__":
    ir = InputReader("input.txt")
    cl = Solver(ir)
    print(cl.part1())
    print(cl.part2())

File: 2024/days/day09/solve.go
package main

import (
	"fmt"
	"os"
	"strings"
)

const (
	sol   = "\033[31;47m"
	reset = "\033[0;0m"
)

type block struct {
	Index, Id, Len int
}

type freeListNode struct {
	indx, len int
	active    bool
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func parseInput(inp []byte) ([]block, []int) {
	blocks := make([]block, len(inp)/2)
	disk := make([]int, 0)
	stripped := strings.Trim(string(inp), "\n")

	block_id := 0
	free_block := false
	for _, s := range stripped {
		blen := int(s - '0')
		to_add := block_id
		if free_block {
			to_add = -1
		} else {
			blocks = append(blocks, block{Index: len(disk), Id: block_id, Len: blen})
			block_id++
		}
		for i := 0; i < blen; i++ {
			disk = append(disk, to_add)
		}
		free_block = !free_block
	}

	return blocks, disk
}

func part1(disk []int) uint64 {
	disklen := len(disk)
	first_free := 0
	for i := disklen - 1; i >= 0; i-- {
		blck := disk[i]

		if blck < 0 {
			continue
		}
		broke := false
		for disk[first_free] > -1 {
			first_free++
			if first_free >= disklen {
				broke = true
				break
			}
		}
		if broke {
			break
		}
		if i <= first_free {
			break
		}
		disk[first_free] = blck
		disk[i] = -1
		continue

	}
	res := uint64(0)
	for i, b := range disk {
		if b < 0 {
			break
		}
		// fmt.Println("{}{}", i, b)
		res += uint64(i * b)
	}

	return res
}

func part2(disk []int, blocks []block) int {
	blockslen := len(blocks)
	free_list := make([]freeListNode, 0)
	free_start := 0
	free_len := 0
	for i, b := range disk {
		if b < 0 {
			if free_len == 0 {
				free_start = i
			}
			free_len++
		} else {
			if free_len > 0 {
				free_list = append(free_list, freeListNode{indx: free_start, len: free_len, active: true})
				free_len = 0
			}
		}
	}
	finish := false
	for j := blockslen - 1; j >= 0; j-- {
		curr_block := blocks[j]
		for k := 0; k < len(free_list); k++ {
			free_node := &free_list[k]
			if !free_node.active {
				continue
			}
			if curr_block.Index <= free_node.indx {
				break
			}
			if free_node.len >= curr_block.Len {
				for i := 0; i < curr_block.Len; i++ {
					disk[curr_block.Index+i] = -1
					disk[free_node.indx+i] = curr_block.Id
				}
				if free_node.len > curr_block.Len {
					free_node.len -= curr_block.Len
					free_node.indx += curr_block.Len
				} else {
					free_node.active = false
				}
				break
			}
		}
		if finish {
			break
		}
	}

	res := 0
	for i, b := range disk {
		if b < 0 {
			continue
		}
		res += i * b
	}
	return res
}

func main() {
	cont, err := os.ReadFile("input.txt")
	check(err)
	_, disk := parseInput(cont)
	disk1 := make([]int, len(disk))

	fmt.Println("Part 1:")
	copy(disk1, disk)
	fmt.Println(part1(disk1))

	// Part 2
	// fmt.Println("===========================================================================")
	// copy(disk1, disk)
	// fmt.Println("Part 2:")
	// fmt.Println(part2(disk1, blocks))

}
